Tutorial on custom IP cores in the FPGA

npetersen2 opened the pull request:
This PR does 3 things:
1. Create an `Advanced Tutorials` section after the existing `Tutorials` section
2. Move the `Custom Control Rate` tutorial to the advanced section
3. Add a new advanced tutorial about IP cores

<img src="https://github.com/Severson-Group/docs.amdc.dev/assets/20168990/2c6604b1-3a40-46b1-a63c-4a8401185935" width="300px" />

npetersen2 commented:
@codecubepi if you get a chance, can you review this PR (see PDF above)?

I am curious if you think everything is explained clearly, things are missing, etc.

In other words, if you only had this tutorial to go by when you started with the FPGA work on the AMDC, would it be enough for you to understand how IP cores work?

npetersen2 commented:
Once @codecubepi reviews, I think we should merge this PR.

I have not implemented this in hardware yet, but let's have the folks starting with the AMDC and FPGA development this summer do this tutorial and gather the experimental results :)

codecubepi commented on source/getting-started/advanced-tutorials/fpga-ip-core/index.md:
Maybe mention CS354 and/or CS400, as C knowledge is required for writing drivers?

npetersen2 replied:
Done

codecubepi commented on source/getting-started/advanced-tutorials/fpga-ip-core/index.md:
Another instance of PS

npetersen2 replied:
Done

codecubepi commented on source/getting-started/advanced-tutorials/fpga-ip-core/index.md:
Why aren't we showing how to get the base address using the define in xparameters.h?

npetersen2 replied:
Eh, it should explain this...

It is somewhat complicated since the `xparameters.h` file is only generated from the SDK side when the `bsp` projects are created from the `hardware_wrapper` project, which itself is created via the `hdf` file which is exported from Vivado.......

I'll just add that:

> Once the SDK environment is set up, all IP core base addresses in the FPGA are available as auto-generated defines in the `xparameters.h` file.

npetersen2 replied:
Done

codecubepi commented on source/getting-started/advanced-tutorials/fpga-ip-core/index.md:
Should be "_is_ quite the feat."

npetersen2 replied:
Done

codecubepi reviewed (commented):
@npetersen2 

Wow, this looks like a really great, thorough tutorial.  I expect this will be a big help onboarding people working on creating custom IP.

Maybe consider linking [your serial LED blog post at the end of the tutorial](https://nathanpetersen.com/2020/07/17/designing-a-firmware-driver-for-serially-addressable-leds-for-xilinx-zynq-7000/)? Granted this new doc goes way further into detail than the blog post, but I think there could still be benefit in exposing new researchers into another task that can be implemented using custom IP.

codecubepi commented:
Oh, one more thing I forgot to note above... would it be good to mention somewhere that the custom IP we create is able to be "Edited in IP Packager", but the IP provided by Vivado/etc is not? I know this was a question I asked you once. 

I had it as a review note in my phone when I started my review the other night, but I didn't have it specifically attached to a certain section of the doc.

npetersen2 commented on source/getting-started/advanced-tutorials/fpga-ip-core/index.md:
Good point, I redirected the text to discuss the PS/PL as an aside to help the reader understand the Xilinx docs, and not a main point. I also changed the rest of this tutorial to use DSP and FPGA instead of PS and PL.

npetersen2 commented on source/getting-started/advanced-tutorials/fpga-ip-core/images/vivado_bd_labeled.png:
Unfortunately, I did not save the source ppt file which created this png, so I'll leave the png as is. :)

Someone else can remake it is desired

npetersen2 commented on source/getting-started/advanced-tutorials/fpga-ip-core/index.md:
Done

npetersen2 commented on source/getting-started/advanced-tutorials/fpga-ip-core/index.md:
Done

npetersen2 commented on source/getting-started/advanced-tutorials/fpga-ip-core/images/vivado_ip_wizard1.png:
;) Some fun photoshop for no actual reason haha

npetersen2 commented on source/getting-started/advanced-tutorials/fpga-ip-core/index.md:
Please!

npetersen2 commented on source/getting-started/advanced-tutorials/fpga-ip-core/index.md:
Yea, I agree.

I wrote the editing IP section as a generic guide, but then the next several sections go over editing the `my_custom_adder` IP with screenshots.

So, I intended this section `5` to be a future reference people could follow, with specific example given below.

I'll add a comment that explains that a bit more in the text.

npetersen2 commented on source/getting-started/advanced-tutorials/fpga-ip-core/index.md:
Done, added a new section `IP Core Inputs and Outputs` which discusses this

npetersen2 commented:
Thanks @codecubepi !

I think I addressed all your feedback, please see the latest commit for a diff.

Also, here is the updated tutorial in PDF form:

[Tutorial Custom FPGA IP Core - AMDC Platform - v2.pdf](https://github.com/Severson-Group/docs.amdc.dev/files/11779686/Tutorial.Custom.FPGA.IP.Core.-.AMDC.Platform.-.v2.pdf)

...this is quite the article! long! we'll have to see how this fares "in the wild" as new people read and complete the steps.

codecubepi commented:
Looks great!

elsevers reviewed (approved):
Excellent -- I did a quick scan of the PDF rendering and this looks very useful. Thanks for pulling it together.
